src/plone.server/plone/server/traversal.py

```diff
@@ -169,60 +191,39 @@ def __init__(self, resource, request, view, rendered):
-                        if SHARED_CONNECTION is False:
-                            txn.abort()
-                        else:
-                            await sync(request)(txn.abort)
+                    if isinstance(view_result, ErrorResponse) or \
+                            isinstance(view_result, UnauthorizedResponse):
+                        # If we don't throw an exception and return an specific
+                        # ErrorReponse just abort
+                        await abort(txn, request)
                     else:
-                        if SHARED_CONNECTION is False:
-                            await txn.acommit()
-                        else:
-                            await sync(request)(txn.commit)
+                        await commit(txn, request)
+
             except Unauthorized as e:
-                eid = uuid.uuid4().hex
-                message = _('Not authorized to render operation') + ' ' + eid
-                logger.error(
-                    message,
-                    exc_info=e)
-                await sync(request)(txn.abort)
-                view_result = UnauthorizedResponse(message)
+                await abort(txn, request)
+                view_result = generate_unauthorized_response(e, request)
             except Exception as e:
-                eid = uuid.uuid4().hex
-                message = _('Error on execution of view') + ' ' + eid
-                logger.error(
-                    message,
-                    exc_info=e)
-                await sync(request)(txn.abort)
-                view_result = ErrorResponse(
-                    'ServiceError',
-                    message
-                )
+                await abort(txn, request)
+                view_result = generate_error_response(
+                    e, request, 'ServiceError')
         else:
             try:
                 view_result = await self.view()
             except Unauthorized as e:
-                eid = uuid.uuid4().hex
-                message = _('Not authorized to render operation') + ' ' + eid
-                logger.error(
-                    message,
-                    exc_info=e)
-                view_result = UnauthorizedResponse(message)
+                view_result = generate_unauthorized_response(e, request)
             except Exception as e:
-                eid = uuid.uuid4().hex
-                message = _('Error on execution of view') + ' ' + eid
-                logger.error(
-                    message,
-                    exc_info=e)
-                view_result = ErrorResponse(
-                    'ViewError',
-                    message)
+                view_result = generate_error_response(e, request, 'ViewError')
+
+        # If we want to close the connection after the request
+        if SHARED_CONNECTION is False and hasattr(request, 'conn'):
+            request.conn.close()
+
+        futures_to_wait = request._futures.values()
```

Comment:
clever! Can we have an api for developers to utilize this feature? Usually, a property that starts with "_" denotes that it should not be publicly used. Could we add a method "add_future" to the request object?